### PR TITLE
Part 1: regex-automata, relative backref, escaping inhibition in classes, Expr::Assertion, Lit { casei }

### PR DESCRIPTION
**Pull request by ruihe774**
This PR contains the first part of #117:

- Switch to use `regex-syntax` and `regex-automata` for anchored delegated search.
- Support `\g` and relative backrefs.
- Inhibit escaping in character classes.
- Implement assertions in VM.
- ~Implement case-insensitive Lit in VM.~

Sorry but I cannot break down it any further. Conflicts would eat me🤣. And a huge number of diff line count is caused by the moving of `Expr` into `parse.rs`, so this is actually a PR with less than 500 lines added. 

Some tests are disabled temporarily, which will be updated and re-enabled in later PR.

### CODE REVIEW

File: src/compile.rs [hunk 1]
```diff
@@ -444,14 +445,22 @@ impl Compiler {
...
-        // TODO: might want to do something similar for case insensitive literals
-        // (have is_literal return an additional bool for casei)
-        if infos.iter().all(|e| e.is_literal()) {
-            let mut val = String::new();
+
+        if infos
+            .iter()
+            .all(|info| matches!(info.expr, Expr::Literal { .. }))
+        {
+            // Parser::optimize() will fold literals, so infos.len() is usually 1.
```
Comment (ruihe774) [inline src/compile.rs:453]: Well, that will reside in another PR...

File: src/parse.rs [hunk 2]
```diff
@@ -38,16 +44,281 @@ const FLAG_SWAP_GREED: u32 = 1 << 3;
 const FLAG_IGNORE_SPACE: u32 = 1 << 4;
 const FLAG_UNICODE: u32 = 1 << 5;
 
-#[cfg(not(feature = "std"))]
-pub(crate) type NamedGroups = alloc::collections::BTreeMap<String, usize>;
-#[cfg(feature = "std")]
-pub(crate) type NamedGroups = std::collections::HashMap<String, usize>;
+/// Regular expression AST. This is public for now but may change.
+#[derive(Debug, Clone, PartialEq, Eq)]
+#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
```
Comment (ruihe774) [inline src/parse.rs:49]: This is actually for serde support (which will be in another PR). Just ignore it and it will not cause any issues for now.

File: src/parse.rs [hunk 3]
```diff
@@ -38,16 +44,281 @@ const FLAG_SWAP_GREED: u32 = 1 << 3;
...
+#[derive(Debug, PartialEq, Eq, Clone, Copy)]
+pub enum Assertion {
+    /// Start of input text
+    StartText,
+    /// End of input text
+    EndText,
+    /// Start of a line
+    StartLine {
+        /// CRLF mode
+        crlf: bool,
```
Comment (ruihe774) [inline src/parse.rs:176]: Not implemented in parser for now.

File: Cargo.lock.direct-minimal-versions
Comment (ruihe774) [inline Cargo.lock.direct-minimal-versions:1]: This is generated using [direct-minimal-versions](https://doc.rust-lang.org/cargo/reference/unstable.html#direct-minimal-versions).

Comment (ruihe774) [inline Cargo.lock.direct-minimal-versions:1]: This is generated using [direct-minimal-versions](https://doc.rust-lang.org/cargo/reference/unstable.html#direct-minimal-versions).

File: Cargo.lock.minimal-versions
Comment (ruihe774) [inline Cargo.lock.minimal-versions:1]: This is generated by [minimal-versions](https://doc.rust-lang.org/cargo/reference/unstable.html#minimal-versions).

File: src/vm.rs [hunk 4]
```diff
@@ -484,12 +487,62 @@ pub(crate) fn run(
...
+                        }
+                    } else {
+                        if !eq_test_exact_size(sb.iter().copied(), val.bytes(), |a, b| {
+                            // Do we need a == b shortcut? Is to_ascii_lowercase() fast?
+                            a == b || a.to_ascii_lowercase() == b.to_ascii_lowercase()
+                        }) {
+                            if !s.get(ix..end).map_or(false, |s| {
+                                eq_test(
+                                    s.chars().flat_map(char::to_lowercase),
+                                    val.chars().flat_map(char::to_lowercase),
```
Comment (robinst) [inline src/vm.rs:537]: Was this copied from somewhere? Case-insensitive matching is a big can of worms. I'm not sure we should try to do it in fancy-regex. I'd prefer that we keep delegating these to regex.

File: tests/oniguruma.rs [hunk 5]
```diff
@@ -220,11 +213,7 @@ fn oniguruma() {
             assert!(result.is_some(),
                     "Expected ignored test to fail, but it succeeded. Remove it from the ignore file: {}", &test.source);
             let failure = result.unwrap();
-            // rustc before version 1.53 unnecessarily escapes slashes in debug format strings
-            // meaning when run on our MSRV, we get different failure text than on the latest stable Rust
-            // so here we try removing the backslash when comparing. After our MSRV is bumped up to
-            // 1.53 or higher, we can remove this part again.
```
Comment (robinst) [inline tests/oniguruma.rs:226]: Nice, thanks.

**Review by robinst — CHANGES_REQUESTED** [review]
Did a first pass, added some comments.

File: src/parse.rs [hunk 6]
```diff
@@ -38,16 +44,279 @@ const FLAG_SWAP_GREED: u32 = 1 << 3;
 const FLAG_IGNORE_SPACE: u32 = 1 << 4;
 const FLAG_UNICODE: u32 = 1 << 5;
 
-#[cfg(not(feature = "std"))]
-pub(crate) type NamedGroups = alloc::collections::BTreeMap<String, usize>;
-#[cfg(feature = "std")]
-pub(crate) type NamedGroups = std::collections::HashMap<String, usize>;
```
[the remark this answers is not in the document]
Comment (robinst) [reply src/parse.rs:44]: Not sure why you marked this as resolved, still looks like the same change?

File: src/parse.rs [hunk 7]
```diff
@@ -839,7 +1134,7 @@ pub(crate) fn parse_id<'a>(s: &'a str, open: &'_ str, close: &'_ str) -> Option<
 }
 
 fn is_id_char(c: char) -> bool {
-    c.is_alphanumeric() || c == '_'
+    c.is_alphanumeric() || c == '_' || c == '-'
```
[the remark this answers is not in the document]
Comment (robinst) [reply src/parse.rs:870]: Can you provide an explanation?

File: src/vm.rs (hunk 4 above)
Comment (robinst) [reply src/vm.rs:537]: I'd prefer not to include this in this PR. It would need a lot of tests and as I said before, I'd rather just delegate this to regex for correctness. If it's in a separate PR we can look at benchmarks too.

**Review by robinst — CHANGES_REQUESTED** [review]
Thanks, much better to review! Added some comment to unresolved parts. Once those are addressed we can hopefully merge this and get to the next one :).

File: src/parse.rs [hunk 8]
```diff
@@ -839,7 +867,7 @@ pub(crate) fn parse_id<'a>(s: &'a str, open: &'_ str, close: &'_ str) -> Option<
...
```
Comment (robinst) [inline src/parse.rs:870]: Note that both Ruby and Python don't allow `-` in group names:

```
SyntaxError ((irb):14: invalid group name <foo-bar>>: /(?<foo-bar>\w+)%\k<foo-bar>/)
```

```
re.error: bad character in group name 'foo-bar' at position 4
```

But I now see that the reason you're doing this change is so that you can parse `\k<-1>`. The code needs to be changed around a bit. Can you add tests for these:

* `\k<foo-bar>` -> invalid group name
* `\k<-1>` -> works as a relative backref
* `\k<+1>` -> works as a relative backref (different from `\k<1>`)

To support that, I'd probably just check the first character within `parse_named_backref`. If it's a `-` or `+`, parse the rest as a number. Otherwise, parse the rest with `c.is_alphanumeric() || c == '_'`.

File: src/vm.rs [hunk 9]
```diff
@@ -484,12 +487,62 @@ pub(crate) fn run(
...
+                }
+                Insn::Lit { ref val, casei } => {
+                    let end = ix.saturating_add(val.len()).min(s.len());
+                    let sb = &s.as_bytes()[ix..end];
+                    if !casei {
+                        if sb != val.as_bytes() {
+                            break 'fail;
+                        }
+                    } else {
+                        if !eq_test_exact_size(sb.iter().copied(), val.bytes(), |a, b| {
```
Comment (robinst) [inline src/vm.rs:530]: You removed the other part of this code (in `visit`) but please also remove this part for now.

File: src/parse.rs [hunk 10]
```diff
@@ -38,10 +41,7 @@ const FLAG_SWAP_GREED: u32 = 1 << 3;
 const FLAG_IGNORE_SPACE: u32 = 1 << 4;
 const FLAG_UNICODE: u32 = 1 << 5;
 
-#[cfg(not(feature = "std"))]
-pub(crate) type NamedGroups = alloc::collections::BTreeMap<String, usize>;
-#[cfg(feature = "std")]
-pub(crate) type NamedGroups = std::collections::HashMap<String, usize>;
+pub(crate) type NamedGroups = BTreeMap<String, usize>;
```
Comment (robinst) [inline src/parse.rs:44]: Please undo this change, as far as I can tell it doesn't have anything to do with the PR.

**Review by robinst — CHANGES_REQUESTED** [review]

**Review by robinst — APPROVED** [review]
@ruihe774 I've pushed a few commits to address some things and fix the build, please have a look. If that all looks good, we can merge :)

**Comment by ruihe774** [comment]
LGTM👍

**Comment by robinst** [comment]
Merged! Thanks, looking forward to the next one. Try to keep it smaller than this please :).

**Comment by robinst** [comment]
@ruihe774 Hey, I was just writing the release notes to release this but I noticed that `\g` is not just another syntax for backrefs, it's subexp calls (see https://www.regular-expressions.info/subroutine.html and https://github.com/kkos/oniguruma/blob/master/doc/RE#L460).

I'll revert the `g` part of the change in a new PR (but keeping the relative change, i.e. `\k<-1>`).

**Comment by ruihe774** [comment]
> @ruihe774 Hey, I was just writing the release notes to release this but I noticed that `\g` is not just another syntax for backrefs, it's subexp calls (see https://www.regular-expressions.info/subroutine.html and https://github.com/kkos/oniguruma/blob/master/doc/RE#L460).
> 
> I'll revert the `g` part of the change in a new PR (but keeping the relative change, i.e. `\k<-1>`).

Interesting. So we need a separated implementation for `\g`. I'd like to look into this latter, maybe.
